### DISCUSSION TITLE
An analogy of Contract#connect method from ethers

### DISCUSSION THREAD

**Discussion by bogdan** [NomicFoundation/hardhat#6089] [Troubleshooting]
I found it hard to have a code like this in tests when testing a multirole permission models on contracts:

``` typescript
    const contract = await viem.getContractAt(
      "MinterRoleMock",
      minterRole.address,
      { client: { wallet: minter } }
    );
```

I tried to abstract it like:

``` typescript
const getContractAt = async <T extends string>(
  name: ContractName<T>,
  contract: {address: Address} | Address,
  wallet: WalletClient
) => {
  const address = typeof contract === 'string' ? contract : contract.address;
  return await viem.getContractAt<T>(
    name,
    address,
    { client: { wallet: wallet } }
  );
}

const contract = await getContractAt("MinterRoleMock", minterRole, minter)
```

But because of typescript magic envolved, didn't find any way to make it work.

Any advice on making the operation of attaching a contract to different wallet making one elegant line of code?

Ideal api I think would be like this:

``` typescript
const connect = async <T extends GetContractReturnType>(
  contract: T,
  wallet: WalletClient,
): Promise<T> => {

}
```

**Comment by kanej** [comment]
@schaable is the issue here that the `WalletClient` type is being imported from `viem` directly and building the abstracting function requires types from `hardhat-viem`?

**Reply by schaable** [reply]
Unfortunately, I don’t think it’s possible to write a wrapper on the user side. You’d need to add a new method to the plugin for the types to work. That said, I don’t believe 
```
await getContractAt("MinterRoleMock", minterRole, minter) 
```
is significantly different from 
```
await getContractAt("MinterRoleMock", minterRole, { client: { wallet: minter }})
```
As for the suggested interface, Viem doesn’t allow switching the client (public or wallet) of a contract; you need to retrieve the contract again. Because of that, `connect(contract, wallet)` would have to call `viem.getContract` again, passing the `contract.abi` and `contract.address`. And figuring out how to return the correct type without knowing the contract name might be tricky to implement.

**Reply by bogdan** [reply]
The `Contract` doesn't store its name. If it would things could get much easier:

``` typescript
const connectWallet = <T extends {address: string, name: string}>(contract: T, wallet: WalletClient): Promise<T> => {
  return await view.getContractAt(contract.name, contract.address, {client: {wallet}}) as T;
}
```

Can you add `name` to `Contract`?

**Reply by schaable** [reply]
When we implemented the plugin, we decided against injecting custom properties into `viem` objects to avoid confusing people. Objects could have different properties depending on whether they come from `viem` or `hardhat-viem`, and I still believe that’s a valid concern. That said, we may revisit this in the future if more people are interested in seeing this feature implemented.

**Reply by bogdan** [reply]
All of that is too smart to be effective 😢 .
